update to pydantic 2

thomas-maschler opened the pull request:
**Related Issue(s):**

- #593 

**Description:**
This PR updates dependencies to Pydantic 2 and Stac-Pydantic 3
It removes the internal Stac, Search and Opertator Types in favor of Types from Stac Pydantic

**PR Checklist:**

- [x] `pre-commit` hooks pass locally
- [x] Tests pass (run `make test`)
- [x] Documentation has been updated to reflect changes, if applicable, and docs build successfully (run `make docs`)
- [x] Changes are added to the [CHANGELOG](https://github.com/stac-utils/stac-fastapi/blob/main/CHANGES.md).

lossyrob commented:
@thomas-maschler , moving the STAC types away from the TypeDict implementations in stac_fastapi.types to Pydantic types could have performance implications. Specifically I'm worried about the STAC types encoding Client output. These were originally moved away from Pydantic models as they tended to be slow for large payloads, e.g. returning an ItemCollection from search that contains 1000 large Sentinel 2 Items. Running Pydantic validation against known valid Items that come out of PgSTAC as part of response causes additional latency without a lot of value. I know the Pydantic 2 backend changes means validation will be faster, but are we introducing the potential for increased latency in the API routes because of the move away from TypedDicts in e.g. BaseCoreClient in this PR? If so, could the upgrade be done for but leave the TypeDict usage as is? Or have you seen that these changes don't introduce additional latency due to the magic of Rust 🪄?

thomas-maschler commented:
I can see your concerns regarding latency. I did not benchmark this myself. Others reported a [speed improvement of x5](https://thedataquarry.com/posts/why-pydantic-v2-matters/#results) between pydantic 1 and 2. I was assuming that this would be enough to address latency concerns.

The benefits I saw in keeping the pydandic models are peace of mind, due to basic checks and setting of default values you wouldn't have to store in the DB. I have a use case, where data are stored in a legacy system and I have to translate them to STAC at runtime. Being able to push part of this into Pydantic makes the job easier.

If latency is still the main concern I could see several routes

1. Keep TypedDicts
Be explicit in the documentation that responses are not type checked and it is the responsibility of the use to do so. 

2. Allow to skip validation for ItemCollections and Collections

```
from pydantic import SkipValidation
from stac_pydantic import api
from typing import Sequence

class ItemCollectionUnsafe(api.ItemCollection):
    features: Sequence[SkipValidation[api.Item]]
 
class CollectionsUnsafe(api.Collections):
    collections: Sequence[SkipValidation[api.Collection]]
```

With minimal code changes, users can use the `Unsafe` subclasses as their return model, telling pydantic to skip validation for items and collections when making bulk requests. 

3. Add some configuration options to let the user choose between current implementation options 1 or 2 above.

rhysrevans3 commented:
Couple of notes from my pull request:
1. If typing `List` and `Dict` are being used [this part](https://github.com/stac-utils/stac-fastapi/pull/625/files#diff-f84457de825c3fa96fa65673a6f85442ac57bd10f9f939cb7411780d3b6fabb5R46-R47) will need to be changed as in it's current form it will miss off the `Optional` typing of nested `Lists` & `Dicts`.
2. The [types.links](https://github.com/stac-utils/stac-fastapi/blob/02f2702a64476ee63bca90a524ce9fc4d2cc91de/stac_fastapi/types/stac_fastapi/types/links.py) seems to be covered by stac-pydantic links and [link_factory](https://github.com/stac-utils/stac-pydantic/blob/main/stac_pydantic/api/utils/link_factory.py).
3. The stac-pydantic Item/Collection/ItemCollection types are not json serialisable so will need to be json encoded/converted to dictionaries before they're returned from the API.

thomas-maschler commented:
@rhysrevans3, regarding your first point, can you give me an example of when this fails? I cannot replicate it. I tried this, where the filter attribute is defined as `Optional[Dict]`

```python

@pytest.mark.parametrize(
    "filter,passes",
    [(None, True), ({"test": "test"}, True), ("test==test", False), ([], False)],
)
def test_create_post_request_model(filter, passes):
    extensions = [FilterExtension()]
    request_model = create_post_request_model(extensions, BaseSearchPostRequest)

    if not passes:
        with pytest.raises(ValidationError):
            model = request_model(filter=filter)
    else:
        model = request_model(
            collections=["test1", "test2"],
            ids=["test1", "test2"],
            bbox=[0, 0, 1, 1],
            datetime="2020-01-01T00:00:00Z",
            limit=10,
            filter=filter,
            **{"filter-crs": "epsg:4326", "filter-lang": "cql2-text"},
        )

        assert model.collections == ["test1", "test2"]
        assert model.filter_crs == "epsg:4326"
        assert model.filter == filter
```

lossyrob commented:
@thomas-maschler understood about the value of validation in the response classes for other use cases.

I think a solution where users could skip validation based on configuration would be great. In my previous testing (a year or so ago) it was the Pydantic validation that was causing slowdowns. I'd still want to validate that there's no other performance hits besides validation for using Pydantic models over TypeDicts before integrating these changes, but I have enough confidence that this would address the performance implications, and any unforeseen issues could be addressed in follow up work.

Would you be willing to add that SkipValidation configuration to this PR?

vincentsarago commented:
about performances, in [**TiPg**](https://github.com/developmentseed/tipg) we also went for the `TypedDict` way because pydantic validation was really slowing down the response (even with pydantic 2.0)

rhysrevans3 commented:
@thomas-maschler I found it was an issue for nested List/Dicts types: for example the `sortby` field in the SortExtensionPostRequest is
```
sortby: Optional[List[PostSortModel]]
```
But when I check the `FieldInfo` shows `required=True`
```
>>> from stac_fastapi.extensions.core.sort.request import SortExtensionPostRequest
>>>SortExtensionPostRequest.model_fields['sortby']
FieldInfo(annotation=Union[List[SortExtension], NoneType], required=True)
```

thomas-maschler commented:
@rhysrevans3 this is b/c we didn't set a default value. Right now the model requires either `None` or a list of `PostSortModel`. If you set the default to `None` the FieldInfo will say `required=False`. However, `None` is filled in when `sortby` is not provided so it still works even without a default value.


Regarding your point 2.
The `type.Links` are no Pydantic Models and are not used anywhere in the package. I prefer to leave them in right now to not break too much. I am not sure if any of the backends currently rely on it.

Regarding your point 3.
All pydantic models are json serializable. either with `.model_dump_json()` (output as `str`) or `.model_dump(mode="json")` (output as `dict`). I added some test, directly returning the pydantic `Item`, `Collection` and `ItemCollection` and this works for me.

I am still working on the switch between `TypeDict` and `Pydantic` response models and will probably push my updates later today.

rhysrevans3 commented:
@thomas-maschler ah that makes more sense thank you for the explanation. I guess this just need to be filled in the backend I'm using. 

Point 2 that sounds like a good idea. But don't the links in the [api.landing_page](https://github.com/stac-utils/stac-pydantic/blob/main/stac_pydantic/api/landing.py) used [here](https://github.com/stac-utils/stac-fastapi/blob/main/stac_fastapi/types/stac_fastapi/types/core.py#L260-L312) need to be stac-pydantic `Links`? I guess with the validation removed this won't be an issue.

Point 3 sorry yes I meant `json.dumps` or `orjson` won't serialise them which I think fastapi uses for returning JSON/GEOJSON content. So was more of a note that the other repos will need to serialise or convert to a `dict` before returning the api routes. I think this effect the [landing pages](https://github.com/stac-utils/stac-fastapi/blob/main/stac_fastapi/types/stac_fastapi/types/core.py#L354-L406). however, I might be missing something obvious.

thomas-maschler commented:
@lossyrob can you take a look at my recent changes, in particular [8118f10](https://github.com/stac-utils/stac-fastapi/pull/625/commits/8118f10be27f04d5ce2143b5fa9dbbd1b4fd92c3).

- In `types.config` I added a parameter `validate_response` that defaults to `False` (same as now).
- I added `types.stac` back in
- I added `types.response_model` which implements the switch
- In `api.core` I made sure we always use response_model classes and added some extra type-checking 
- I added additional tests to make sure this works correctly, in particular, `test_response_model` and `test_app`

I stumbled over some other bugs regarding field aliases in the filter extension that I didn't manage to fix. I will file an issue for that.
I also noticed that there are still some pydantic extension models that are either out of sync with stac-pydantic or not present in stac-pydantic. I will port them over and open another PR once they make their way through the system.

thomas-maschler commented:
@rhysrevans3 
regarding your point 3: please look at `test_app.py` in this PR. Serialization works just fine

rhysrevans3 commented:
@thomas-maschler sorry about that I've been getting `TypeError: Object of type LandingPage is not JSON serializable` on the landing page when trying to update the [stac-fastapi-elasticsearch-opensearch](https://github.com/stac-utils/stac-fastapi-elasticsearch-opensearch) to use this branch and assumed it was because of the switch to `api.LandingPage` type but this is my first time using pydantic models. I'll keep digging. Thanks for all your help!

thomas-maschler commented:
I haven't looked at the backends. We use a custom one.
Try importing `types.response_model.LandingPage` and see if it works now.

If you want to use the Pydantic models set the envvar `VALIDATE_RESPONSE=True`

vincentsarago commented:
Hi everyone, I'm going to take over the review. 

@thomas-maschler did you had anything to add or is this PR ready?

thomas-maschler commented:
Thanks, this is ready. The most controversial part is probably the response model

jonhealy1 commented:
I added a comment about adding something in the readme explaining the option to choose the response model.

vincentsarago commented on stac_fastapi/api/tests/test_app.py:
```diff
@@ -0,0 +1,149 @@
+    ],
+)
+def test_app(validate, response_type, collection_dict, item_dict, cleanup):
+    os.environ["VALIDATE_RESPONSE"] = str(validate)
```
could we use monkeypatch here instead of writing the env variable 
https://docs.pytest.org/en/6.2.x/monkeypatch.html

vincentsarago commented on stac_fastapi/api/tests/test_app.py:
```diff
@@ -0,0 +1,149 @@
+        ) -> response_model.ItemCollection:
+            # FIXME: hyphen alias for filter_crs and filter_lang are currently not working
+            # assert kwargs.get("filter_crs") == "EPSG:4326"
+            # assert kwargs.get("filter_lang") == "cql-test"
```
I'm not sure to get this. Did you open an issue about it yet?

thomas-maschler replied:
yep, https://github.com/stac-utils/stac-fastapi/issues/638

vincentsarago commented on stac_fastapi/api/tests/test_app.py:
```diff
@@ -0,0 +1,149 @@
+import importlib
```
can we add some docstrings explaining what this `test_app` will test?

vincentsarago commented on stac_fastapi/api/setup.py:
```diff
@@ -7,8 +7,8 @@
-    "pydantic[dotenv]<2",
-    "stac_pydantic==2.0.*",
+    "pydantic[dotenv]>=2",
+    "stac_pydantic>=3",
```
we should maybe restrict stac_pydantic to `<4.0`

vincentsarago replied:
maybe we should just remove the requirements here and rely on `stac-fastapi.types` requirements?

maybe we should also pin `stac-fastapi.types` version

thomas-maschler replied:
Following [Henry Schreiner advice](https://iscinumpy.dev/post/bound-version-constraints/) I usually only add lower bounds.

vincentsarago commented on stac_fastapi/api/setup.py:
```diff
@@ -7,8 +7,8 @@
     "attrs",
-    "pydantic[dotenv]<2",
-    "stac_pydantic==2.0.*",
+    "pydantic[dotenv]>=2",
```
I think `dotenv` was removed in pydantic 2.0  

python-dotenv is now the dependency of pedantic-settings

vincentsarago commented on stac_fastapi/api/stac_fastapi/api/models.py:
```diff
@@ -44,40 +43,19 @@ def create_request_model(
-                )
-                fields[k] = (v.outer_type_, body)
+            for k, field_info in model.model_fields.items():
+                fields[k] = (field_info.annotation, field_info)
```
can you elaborate why we changed this?

thomas-maschler replied:
I am not sure why we used fastapi.Body in first place, but some API changes in pydantic 2 didn't allow to do a 1:1 mapping between `FieldInfo` and `Body`. However, you can now pass a `dict` of `FieldInfo` objects to `create_model` and it will populate the model for you. This is much cleaner than the previous solution.

vincentsarago commented on stac_fastapi/api/tests/test_models.py:
```diff
@@ -0,0 +1,100 @@
+        limit=10,
+        filter="test==test",
+        # FIXME: hyphen aliases are not properly working
+        # **{"filter-crs": "epsg:4326", "filter-lang": "cql2-text"},
```
do we know if this was working before pydantic 2.0 ?

vincentsarago replied:
maybe this is linked to the changes at https://github.com/stac-utils/stac-fastapi/pull/625/files#diff-f84457de825c3fa96fa65673a6f85442ac57bd10f9f939cb7411780d3b6fabb5R47

thomas-maschler replied:
hmm, let me test that

thomas-maschler replied:
I tested locally and reverting back to using the fastapi.Body instead of FieldInfo doesn't change that.
This section only handles POST requests that use the Pydantic BaseModel.

The issue with aliases only appears with GET requests that use attrs classes with Pydantic Fields. I don't think that ever worked. I will do some more testing tomorrow.

thomas-maschler replied:
I did some more testing and I am confident that this bug already existed before my changes and I only discovered it. I added several tests and linked them back to the issue I created for the bug.

vincentsarago commented on stac_fastapi/extensions/setup.py:
```diff
@@ -7,8 +7,8 @@
-    "pydantic[dotenv]<2",
-    "stac_pydantic==2.0.*",
+    "pydantic[dotenv]>=2",
+    "stac_pydantic>=3",
```
same for `api` 
- `dotenv` is not an option anymore
- limit `stac_pydantic` upper version
- pin `stac-fastapi.types` and `stac-fastapi.api` version

lossyrob commented:
> @lossyrob can you take a look at my recent changes, in particular [8118f10](https://github.com/stac-utils/stac-fastapi/pull/625/commits/8118f10be27f04d5ce2143b5fa9dbbd1b4fd92c3).
> 
> * In `types.config` I added a parameter `validate_response` that defaults to `False` (same as now).
> * I added `types.stac` back in
> * I added `types.response_model` which implements the switch
> * In `api.core` I made sure we always use response_model classes and added some extra type-checking
> * I added additional tests to make sure this works correctly, in particular, `test_response_model` and `test_app`
> 
> I stumbled over some other bugs regarding field aliases in the filter extension that I didn't manage to fix. I will file an issue for that. I also noticed that there are still some pydantic extension models that are either out of sync with stac-pydantic or not present in stac-pydantic. I will port them over and open another PR once they make their way through the system.

Thanks for this change, this for sure satisfies the concerns I brought up.

Looking forward to this making its way into a release soon!

vincentsarago commented on stac_fastapi/api/setup.py:
```diff
@@ -1,16 +1,20 @@
+ver_path = convert_path("stac_fastapi/api/version.py")
+with open(ver_path) as ver_file:
+    exec(ver_file.read(), main_ns)
+
```
Can we avoid doing this? 

I think it's fine to manually hard code the version here. 

We can then use something like `bump2version` to update all the files we want when doing a release

e.g. https://github.com/developmentseed/titiler/blob/main/.bumpversion.cfg

thomas-maschler replied:
ok, i will hard code it for now.
I created a follow up ticket
https://github.com/stac-utils/stac-fastapi/issues/648

vincentsarago commented on stac_fastapi/api/tests/test_app.py:
```diff
@@ -30,8 +33,9 @@ def cleanup():
-def test_app(validate, response_type, collection_dict, item_dict, cleanup):
-    os.environ["VALIDATE_RESPONSE"] = str(validate)
+def test_app(validate, response_type, collection_dict, item_dict, cleanup, monkeypatch):
+    monkeypatch.setenv("VALIDATE_RESPONSE", validate)
```
do we still need the `cleanup`?

vincentsarago commented on stac_fastapi/types/stac_fastapi/types/response_model.py:
```diff
@@ -0,0 +1,26 @@
+Collections = response_model.Collections
+Item = response_model.Item
+ItemCollection = response_model.ItemCollection
+try:
+    Conformance = response_model.Conformance
+except AttributeError:
+    # TODO: class name needs to be fixed in stac_pydantic
+    Conformance = response_model.ConformanceClasses
```
ref: https://github.com/stac-utils/stac-pydantic/pull/136

vincentsarago commented on stac_fastapi/types/stac_fastapi/types/search.py:
```diff
@@ -1,79 +1,23 @@
+from pydantic import PositiveInt
+from pydantic.functional_validators import AfterValidator
+from stac_pydantic.api import Search
+from typing_extensions import Annotated
```
There is a lot of changes in this file, can you explain?

thomas-maschler replied:
I removed classes that are now implemented in stac_pydantic

vincentsarago commented on stac_fastapi/types/stac_fastapi/types/core.py:
```diff
@@ -306,10 +309,20 @@ def _landing_page(
                 },
+                {
+                    "rel": Relations.service_desc.value,
+                    "type": MimeTypes.geojson,
```
should this be 
```
"type": MimeTypes.geojson.value,
```

vincentsarago commented on stac_fastapi/types/stac_fastapi/types/core.py:
```diff
@@ -262,8 +265,8 @@ def _landing_page(
         extension_schemas: List[str],
-    ) -> stac_types.LandingPage:
-        landing_page = stac_types.LandingPage(
+    ) -> Dict[str, Any]:
```
Why is this method returning a Dict instead of `response_model.LandingPage` ?

thomas-maschler replied:
This is an internal method and we later modify the dictionary further. Instead of adding toggles between the two response types throughout the code, it is easier to default to a `dict` here and do a final cast when returning the response. Returning a `stac.LandingPage` would also work if that adds more clarity

vincentsarago commented:
Thanks a ton @thomas-maschler

This PR is `massive` and has many unrelated changes to pydantic update but it's too late to refactor so we'll just try to merge this. 

I've posted some comment directly in the files themself, but wanted to ask one `general` question. 

I see that we now have an option for the Model to be either TypeDict or Pydantic Model. This model is used  at two levels (if I understand well):

1. between the Backend and the Client (https://github.com/thomas-maschler/stac-fastapi/blob/pydantic2/stac_fastapi/types/stac_fastapi/types/stac.py#L490)

2. between the Client and the User (via FastAPI response) (https://github.com/thomas-maschler/stac-fastapi/blob/pydantic2/stac_fastapi/api/stac_fastapi/api/app.py#L238-L240) 

Question:
- I wonder if there is a real need to the Pydantic Model to be used at level 1, we could still get validation error when FastAPI would try to validate the response? 
- in https://github.com/thomas-maschler/stac-fastapi/blob/e24351ad2c182a790128d11423cc05870df7dd9f/stac_fastapi/api/stac_fastapi/api/app.py#L238-L240
if `self.settings.enable_response_models` is set to `False` we won't have OpenAPI schema hints for the response. In `TiPG` we overcome this by passing the model directly to the OpenAPI like https://github.com/developmentseed/tipg/blob/main/tipg/factory.py#L688 . Do you think we could do the same here (we can do it in another PR)

thomas-maschler commented:
Oh, I somehow missed out on option 2. That makes things a lot clearer now. 
Let's go with that approach only. It is less convoluted than what I did. 

I will revert my introduced changes and make sure FastAPI uses the correct models.
Let me also see how I can inject the model to OpenAPI

thomas-maschler commented:
ok, I got this reverted but I don't get my tests to pass. Am I doing something wrong?
Even if I set [`enable_response_models` to `True`](https://github.com/stac-utils/stac-fastapi/pull/625/commits/eea9c8227705a35425b7e2f864946a787f9ecaef#diff-654b52c9fb869f6f11556091a7f2814122d2c88ec65453d168a85c26516da5edR81) I get a `200` response with an invalid item.

vincentsarago commented on stac_fastapi/types/stac_fastapi/types/core.py:
```diff
@@ -383,10 +379,8 @@ def landing_page(self, **kwargs) -> response_model.LandingPage:
-            collections = _collections.model_dump(mode="json")
-        else:
-            collections = _collections
+        collections = _collections
```
I don't think we need to create a new variable here! 

```python
collections = self.all_collections(request=kwargs["request"])
for collection in collections["collections"]:
```

vincentsarago commented:
> ok, I got this reverted but I don't get my tests to pass. Am I doing something wrong?
Even if I set [enable_response_models to True](https://github.com/stac-utils/stac-fastapi/pull/625/commits/eea9c8227705a35425b7e2f864946a787f9ecaef#diff-654b52c9fb869f6f11556091a7f2814122d2c88ec65453d168a85c26516da5edR81) I get a 200 response with an invalid item.

🤯 this is so weird!

vincentsarago commented:
```python

import pytest

from pydantic import BaseModel
from fastapi import FastAPI
from fastapi.exceptions import ResponseValidationError
from fastapi.testclient import TestClient


class model(BaseModel):
    value: int


def test_model():
    app = FastAPI()

    # should return 200
    @app.get(
        "/yo",
        response_model=model,
        responses={
            200: {
                "content": {
                    "application/json": {},
                },
                "model": model,
            },
        },
    )
    def yo():
        return {"value": 1}

    # response model
    # so validation should fail
    @app.get(
        "/ye",
        response_model=model,
        responses={
            200: {
                "content": {
                    "application/json": {},
                },
                "model": model,
            },
        },
    )
    def ye():
        return {"value": "b"}

    # No response model
    # so No validation
    # should return 200
    @app.get(
        "/ya",
        responses={
            200: {
                "content": {
                    "application/json": {},
                },
                "model": model,
            },
        },
    )
    def ya():
        return {"value": "b"}

    # SAME With `add_api_route` method
    def yi():
        return {"value": "b"}

    app.router.add_api_route(
        name="yi",
        path="/yi",
        response_model=model,
        responses={
            200: {
                "content": {
                    "application/json": {},
                },
                "model": model,
            },
        },
        methods=["GET"],
        endpoint=yi,
    )

    with TestClient(app) as client:
        resp = client.get("/yo")
        assert resp.status_code == 200

        with pytest.raises(ResponseValidationError):
            client.get("/ye")

        resp = client.get("/ya")
        assert resp.status_code == 200

        with pytest.raises(ResponseValidationError):
            client.get("/yi")
```

I've tried to reproduce this with a simple APP but without success 😬 , ☝️ pass the tests

vincentsarago commented:
ahhhhhh it's because the endpoint response get's automatically wrapped in a `Response` so it passthrough the validation 😭 

https://github.com/tiangolo/fastapi/blob/7e161b3f9e7ae5b6a37ab8ea13b6dd2c1c11eff9/fastapi/routing.py#L281-L284

so 👇 won't raise an issue 
```
    @app.get(
        "/yu",
        response_model=model,
        responses={
            200: {
                "content": {
                    "application/json": {},
                },
                "model": model,
            },
        },
    )
    def yu():
        return JSONResponse({"value": "b"})
```

vincentsarago commented:
I'm pretty we can remove this 

https://github.com/thomas-maschler/stac-fastapi/blob/d0762eb20dfba437376c9a8879524cac0ee9ce7e/stac_fastapi/api/stac_fastapi/api/routes.py#L26

because FastAPI will automatically do it, I'll make a PR on top of your PR

vincentsarago commented on stac_fastapi/extensions/stac_fastapi/extensions/core/context.py:
```diff
@@ -1,4 +1,5 @@
 """Context extension."""
```
The context extension is deprecated! I think we can 🔪 it!

vincentsarago replied:
maybe in another PR

thomas-maschler replied:
https://github.com/stac-utils/stac-fastapi/issues/649

thomas-maschler commented:
Phew, that was a beast. It kind of grew out of hand. Thanks for the help.
One thing left is updating the documentation. I believe, that once I turn that stone, new things will pop up and I prefer to make this a separate PR.

vincentsarago commented:
thanks a lot @thomas-maschler. 

One thing that I want to do before we merge is make sure the performance is 👍, I'll try to set up some benchmark over the weekend

jonhealy1 commented on stac_fastapi/types/stac_fastapi/types/core.py:
```diff
@@ -437,15 +446,11 @@ def get_search(
+        intersects: Optional[str] = None,
         datetime: Optional[Union[str, datetime]] = None,
         limit: Optional[int] = 10,
-        query: Optional[str] = None,
-        token: Optional[str] = None,
-        fields: Optional[List[str]] = None,
-        sortby: Optional[str] = None,
```
What are the reasons for removing these fields?

thomas-maschler replied:
They are used for extensions only and should not be part of this abstract method. Users can implement them in their own subclass if they need to.

jonhealy1 replied:
Makes sense. Thank you.

jonhealy1 commented on stac_fastapi/types/stac_fastapi/types/core.py:
```diff
@@ -633,15 +640,11 @@ async def get_search(
+        intersects: Optional[str] = None,
         datetime: Optional[Union[str, datetime]] = None,
         limit: Optional[int] = 10,
-        query: Optional[str] = None,
-        token: Optional[str] = None,
-        fields: Optional[List[str]] = None,
-        sortby: Optional[str] = None,
```
Same question here.

thomas-maschler replied:
Same reason as above

jonhealy1 commented on stac_fastapi/types/stac_fastapi/types/config.py:
```diff
@@ -30,11 +31,7 @@ class ApiSettings(BaseSettings):
-
-        extra = "allow"
-        env_file = ".env"
+    model_config = SettingsConfigDict(env_file=".env", extra="allow")
```
I think it would be nice to have a small section in the readme explaining how to set the response validation.

thomas-maschler replied:
I added some language to the Readme

jonhealy1 replied:
Looks great. Thank you

jonhealy1 reviewed (changes requested):
Everything looks really good to me. A couple questions and a request to add an explanation about the `enable_response_models` option to the readme. Also interested in any performance testing @vincentsarago is doing.

drnextgis commented:
Do you think [this](https://github.com/stac-utils/stac-fastapi/issues/461) issue is going to be fixed once this PR is merged?

vincentsarago commented:
@drnextgis Hey sorry for not seeing your ticket before, but yes it will be fixed. I stumbled on this to https://github.com/thomas-maschler/stac-fastapi/pull/2

I'll fix also for ~2.4 version of stac-fastapi

vincentsarago commented:
```
----------------------------------- benchmark 'Items': 7 tests -----------------------------------
Name (time in ms)           Min                Max               Mean             Median          
--------------------------------------------------------------------------------------------------
Items Limit: (1)         1.0452 (1.0)       2.1031 (1.0)       1.1329 (1.0)       1.1137 (1.0)    
Items Limit: (10)        1.7471 (1.67)      3.0766 (1.46)      1.8286 (1.61)      1.8087 (1.62)   
Items Limit: (50)        4.5001 (4.31)     64.0863 (30.47)     4.9263 (4.35)      4.6144 (4.14)   
Items Limit: (100)       8.0186 (7.67)     64.3674 (30.61)     8.6762 (7.66)      8.1762 (7.34)   
Items Limit: (200)      14.8327 (14.19)    17.0791 (8.12)     15.1425 (13.37)    15.0344 (13.50)  
Items Limit: (250)      18.1781 (17.39)    58.4482 (27.79)    19.1771 (16.93)    18.4300 (16.55)  
Items Limit: (1000)     69.1144 (66.13)    69.9298 (33.25)    69.6401 (61.47)    69.6901 (62.58)  
--------------------------------------------------------------------------------------------------

-------------------------------- benchmark 'Items With Model validation': 7 tests --------------------------------
Name (time in ms)                          Min                 Max               Mean             Median          
------------------------------------------------------------------------------------------------------------------
Items With Model validation (1)         1.1480 (1.0)        2.3425 (1.0)       1.2314 (1.0)       1.2174 (1.0)    
Items With Model validation (10)        1.7948 (1.56)       2.8828 (1.23)      1.9552 (1.59)      1.9323 (1.59)   
Items With Model validation (50)        4.1826 (3.64)      38.1875 (16.30)     4.7179 (3.83)      4.2844 (3.52)   
Items With Model validation (100)       7.2982 (6.36)      43.5653 (18.60)     8.6554 (7.03)      7.4836 (6.15)   
Items With Model validation (200)      12.8544 (11.20)     47.9423 (20.47)    17.3208 (14.07)    13.2296 (10.87)  
Items With Model validation (250)      16.0176 (13.95)     53.0070 (22.63)    21.2084 (17.22)    16.2596 (13.36)  
Items With Model validation (1000)     61.4995 (53.57)    106.9911 (45.67)    86.3199 (70.10)    99.1915 (81.48)  
------------------------------------------------------------------------------------------------------------------
```

Well, either the validation/serialization is not working or there is no difference 🤯

jonhealy1 commented:
@vincentsarago I thought we had results [here](https://github.com/stac-utils/stac-fastapi/pull/650#issuecomment-2042918120) that seemed to make sense?

thomas-maschler commented:
I believe we now test for correct validation.

vincentsarago commented:
> I believe we now test for correct validation.

@thomas-maschler Yes

> @vincentsarago I thought we had results https://github.com/stac-utils/stac-fastapi/pull/650#issuecomment-2042918120 that seemed to make sense?

@jonhealy1 absolutely, so this mean that validation does not impact the performance 🤯. I would still keep this optional but it might be great to mention this somewhere

thomas-maschler commented:
Ok, I'll update the readme

vincentsarago commented:
thanks @thomas-maschler I was going update this branch tomorrow morning 🙏 

let see if we can get this merged before the end of the week 😄

jonhealy1 commented on stac_fastapi/types/stac_fastapi/types/version.py:
```diff
@@ -1,2 +1,3 @@
 """Library version."""
+
 __version__ = "2.5.4"
```
2.5.5

jonhealy1 reviewed (changes requested):
Nice work. The readme looks great - just a couple of spelling things

vincentsarago commented on stac_fastapi/extensions/setup.py:
```diff
@@ -1,16 +1,14 @@
-    "stac-fastapi.types",
-    "stac-fastapi.api",
+    "stac-fastapi.types==2.4.9",
+    "stac-fastapi.api==2.4.9",
```
not really because we will do #648 ;-)

jonhealy1 replied:
The thing is though it's sort of complicated because let's say we put types to 3.0.0 but types has not been release to pypi yet so the tests don't pass.

jonhealy1 replied:
In sfeos I have to push a new core version manually to pypi before I can push es and os updates with new versions

vincentsarago replied:
we won't change the version before the release

vincentsarago replied:
this is how we do in titiler and it works fine. 

right now the versioning is quite dangerous because if you have multiple version of the modules you could hit some issues. Because the 3 modules are supper linked to each other, the versions should always be the same

jonhealy1 replied:
the release is triggered by a github workflow - if we put stac-fastapi.types==3.0.0 in stac-fastapi.api then the tests will fail so how can we merge?

jonhealy1 replied:
if you don't change the version before the release then the release will point to old versions

jonhealy1 replied:
I don't think it's dangerous becuase stac-fastapi is run in a container or in the cloud - no one should have old versions installed that won't be re-installed

jonhealy1 replied:
Does anyone use them separately?

jonhealy1 replied:
They all have to be on the same version together at the same time pretty much

vincentsarago replied:
ok let's discuss this in an issue 👍

jonhealy1 replied:
See what people think

vincentsarago commented on stac_fastapi/api/setup.py:
```diff
@@ -57,4 +54,5 @@
     install_requires=install_requires,
     tests_require=extra_reqs["dev"],
     extras_require=extra_reqs,
+    version="2.4.9",
```
yeah we should remove this

jonhealy1 reviewed (approved):
Nice work!

vincentsarago reviewed (approved):
🚀 thanks @thomas-maschler

Notes:
- we should wait for the next stac-pydantic release with https://github.com/stac-utils/stac-pydantic/pull/131
- I've created  a `maint-2.x` branch (https://github.com/stac-utils/stac-fastapi/tree/maint-2.x) if we need to continue updating `old` stac-fastapi

jonhealy1 commented:
@vincentsarago when you say wait do you mean wait to merge?

vincentsarago commented:
@jonhealy1 sorry I meant, wait for the 3.0 release :-) 

I'll merge this now 🥳

vincentsarago commented:
🤯 😍 
<img width="1569" alt="Screenshot 2024-04-26 at 9 27 17 AM" src="https://github.com/stac-utils/stac-fastapi/assets/10407788/bef8c3e1-646f-4799-835f-b07378608c99">

<img width="1686" alt="Screenshot 2024-04-26 at 9 27 15 AM" src="https://github.com/stac-utils/stac-fastapi/assets/10407788/47996547-43ac-4466-b174-3ce0830cb380">
